gslib/tests/test_context_config.py

```diff
@@ -0,0 +1,326 @@
+SOMECOMMENT
+"""
+CERT_KEY_WITH_COMMENT_IN_BETWEEN = """-----BEGIN CERTIFICATE-----
+LKJHLSDJKFHLEUIORWUYERWEHJHL
+KLJHGFDLSJKH(@#*&$)@*#KJHFLKJDSFSD
+-----END CERTIFICATE-----
+SOMECOMMENT
+-----BEGIN ENCRYPTED PRIVATE KEY-----
+LKJWE:RUWEORIU)(#*&$@(#$KJHLKDJHF(I*F@YLFHSLDKJFS
+-----END ENCRYPTED PRIVATE KEY-----
+"""
+
+OPEN_TO_PATCH = '__builtin__.open' if six.PY2 else 'builtins.open'
+
+
+@testcase.integration_testcase.SkipForS3('mTLS only runs on GCS JSON API.')
+@testcase.integration_testcase.SkipForXML('mTLS only runs on GCS JSON API.')
+class TestPemFileParser(testcase.GsUtilUnitTestCase):
+  """Test PEM-format certificate parsing for mTLS."""
+
+  def testPemFileWithCommentAtBegin(self):
+    sections = context_config._SplitPemIntoSections(
+        CERT_KEY_WITH_COMMENT_AT_BEGIN, self.logger)
+    self.assertEqual(sections['CERTIFICATE'], CERT_SECTION)
+    self.assertEqual(sections['ENCRYPTED PRIVATE KEY'], KEY_SECTION)
+
+  def testPemFileWithCommentAtEnd(self):
+    sections = context_config._SplitPemIntoSections(
+        CERT_KEY_WITH_COMMENT_AT_END, self.logger)
+    self.assertEqual(sections['CERTIFICATE'], CERT_SECTION)
+    self.assertEqual(sections['ENCRYPTED PRIVATE KEY'], KEY_SECTION)
+
+  def testPemFileWithCommentInBetween(self):
+    sections = context_config._SplitPemIntoSections(
+        CERT_KEY_WITH_COMMENT_IN_BETWEEN, self.logger)
+    self.assertEqual(sections['CERTIFICATE'], CERT_SECTION)
+    self.assertEqual(sections['ENCRYPTED PRIVATE KEY'], KEY_SECTION)
+
+  def testPemFileWithBadFormatEmbeddedSection(self):
+    sections = context_config._SplitPemIntoSections(
+        BAD_CERT_KEY_EMBEDDED_SECTION, self.logger)
+    self.assertIsNone(sections.get('CERTIFICATE'))
+    self.assertEqual(sections.get('ENCRYPTED PRIVATE KEY'), KEY_SECTION)
+
+  def testPemFileWithBadFormatMissingEnd(self):
+    sections = context_config._SplitPemIntoSections(BAD_CERT_KEY_MISSING_END,
+                                                    self.logger)
+    self.assertEqual(sections.get('CERTIFICATE'), CERT_SECTION)
+    self.assertIsNone(sections.get('ENCRYPTED PRIVATE KEY'))
+
+  def testPemFileWithBadFormatMissingBegin(self):
+    sections = context_config._SplitPemIntoSections(BAD_CERT_KEY_MISSING_BEGIN,
+                                                    self.logger)
+    self.assertIsNone(sections.get('CERTIFICATE'))
+    self.assertEqual(sections.get('ENCRYPTED PRIVATE KEY'), KEY_SECTION)
+
+  def testPemFileWithBadFormatMismatch(self):
+    sections = context_config._SplitPemIntoSections(BAD_CERT_KEY_MISMATCH,
+                                                    self.logger)
+    self.assertIsNone(sections.get('CERTIFICATE'))
+    self.assertIsNone(sections.get('ENCRYPTED PRIVATE KEY'))
+
+
+@testcase.integration_testcase.SkipForS3('mTLS only runs on GCS JSON API.')
+@testcase.integration_testcase.SkipForXML('mTLS only runs on GCS JSON API.')
+class TestContextConfig(testcase.GsUtilIntegrationTestCase):
```

Comment:
Should be UnitTest?

Reply:
The last test in this class "feels" like an integration test, but unit test inheritance also works.
Might as well use GsUtilUnitTestCase, assuming it's lighter weight.